INT-1577 Support Stored Credit Card (vaulting) for Stripe V3

clopezh opened the pull request:
[INT-1577](https://jira.bigcommerce.com/browse/INT-1577)
## What?
Add support for Stripe V3 + Vault that can handle stripe validations with stripe.js (createPaymentMethod & handleCardPayment actions)

## Why?
To support Vaulting on Stripe V3 is needed to modify the existing strategy that use Stripe.js elements and actions to handle 3DS flow and confirm a payment intent that previously was created from back-end (which used a new end point created on BigPay). Also a change into ng-checkout was required to support instruments into `HostedWidgetPaymentMethod`

## Testing / Proof
![image](https://user-images.githubusercontent.com/35146660/61543279-c7cbe700-aa08-11e9-8f80-207d22b7514e.png)

![image](https://user-images.githubusercontent.com/35146660/61542964-2fcdfd80-aa08-11e9-8062-f797e0e6ef1b.png)

![image](https://user-images.githubusercontent.com/35146660/61544124-5b51e780-aa0a-11e9-9e45-2cd17524ff0a.png)



![ezgif com-video-to-gif](https://user-images.githubusercontent.com/35146660/60542794-61418c00-9cda-11e9-9d03-4c2cc3976de1.gif)

## Sibling PRs
- [BigPay-Client-JS](https://github.com/bigcommerce/bigpay-client-js/pull/79)
- [NGCheckout](https://github.com/bigcommerce-labs/ng-checkout/pull/1282)
- [BigPay](https://github.com/bigcommerce/bigpay/pull/1730)

@bigcommerce/checkout @bigcommerce/intersys-integrations

metalvegetarianoprogresivo commented on src/payment/strategies/stripev3/stripev3-payment-strategy.spec.ts:
```diff
@@ -24,15 +26,20 @@ import {
 } from '../../../payment';
-import { PaymentArgumentInvalidError } from '../../errors';
+import { getShippingAddress } from '../../../shipping/shipping-addresses.mock';
+import {PaymentArgumentInvalidError} from '../../errors';
```
Add spaces between the curly bracket.
```suggestion
import { PaymentArgumentInvalidError } from '../../errors';
```

clopezh commented on src/payment/payment-action-creator.ts:
```diff
@@ -60,6 +64,23 @@ export default class PaymentActionCreator {
         };
     }
 
+    generatePaymentIntent(
```
Reason behind why we added this new endpoint is because existing one handles ClientToken only, on this one we handle PaymentIntents which maps a different set of data (CustomerId, BillingAddress, CurrencyCode, GrandTotal, etc.), on the other hand, we designed this endpoint to communicate directly SDK and BigPay because we there’s no need to store the token in memory.

clopezh commented on src/order/order-request-body.ts:
```diff
@@ -18,6 +18,12 @@ export default interface OrderRequestBody {
+     * If true, apply to save the credit card as an instrument. It only works if a hosted field needs
+     * to be vaulted.
+     */
+    shouldSaveInstrument?: boolean;
```
`paymentData` type is of 1 of 2 subtypes `CreditCardInstument` or `VaultedInstrument`, our case is neither one. If we decide to handle ours as a `CreditCardInstrument` we still have to pass the rest of cc info which we don’t have access as this is handled by Stripe library. That is why we had to create this new property as part of `OrderRequestBody`.
We believe it is a better place to have this property in here just like `useStoreCredit` flag.

clopezh commented:
Hey @davidchin I addressed some of your comments and also I answer some of your questions, Could you please take a look? ♻️ Thank you!!

davidchin commented on src/payment/strategies/stripev3/stripev3-payment-strategy.ts:
```diff
@@ -70,39 +72,58 @@ export default class StripeV3PaymentStrategy implements PaymentStrategy {
+                        };
+
+                        return this._getStripeJs().handleCardPayment(paymentIntent, stripeCardPaymentOptions)
+                            .then(stripeResponse => {
```
This level of nesting can be removed by chaining with the outer promise instead. i.e.:

```ts
return this._getStripeJs().createPaymentMethod('card', this._cardElement, this._mapStripePaymentMethodOptions())
    .then(stripePaymentMethod => {
        // Do what you have to do with `stripePaymentMethod`
        return this._getStripeJs().handleCardPayment(paymentIntent, stripeCardPaymentOptions);
    })
    .then(stripeResponse => {
        // Do what you have to do with `stripePaymentMethod`
    });
```

clopezh commented:
Hey @davidchin I already addressed the last comments, and also I added extra unit testing to increase the percentage of coverage, Could you please take a look and if possible approve it? Thank you! ♻️

davidchin commented on src/payment/payment-request-transformer.ts:
```diff
@@ -0,0 +1,90 @@
+import PaymentRequestBody from './payment-request-body';
+
+export default class PaymentRequestTransformer {
+    static transform(payment: Payment, checkoutState: InternalCheckoutSelectors): PaymentRequestBody {
```
We prefer not to use `static` methods as it's not possible to mock them. Can you make this an instance method and inject the instance into places where you need it?

clopezh commented:
@davidchin Comment addressed!! 👍 Could you please take a look and approve it? ♻️
